### PR TITLE
Add read/write instrumentation with LTTng

### PR DESCRIPTION
**Pull request by christophebedard**
This adds tracing instrumentation using LTTng for:

* writer creation
* reader creation
* writing
* reading

When a writer or reader entity is created, we note:

* the pointer to the underlying writer/reader
* the topic name
* its GUID

When we write or read, we note:

* the pointer to the underlying writer/reader
* the pointer to the message buffer
* (write only) the source timestamp

This allows us to:

* match a writer or reader with an encapsulating/higher-level object in an application layer (above) using the GUID
* match writer/reader creation events with write/read events using the writer/reader pointers
* match the write/read message buffer pointers with other write/read/publish/take tracepoints in the application level
* match the source timestamp between writing and receiving

Also, while it has to be separate from the ROS 2 instrumentation, these tracepoints can easily be enabled by the tracing-related launch/CLI utilities in https://gitlab.com/ros-tracing/ros2_tracing

See the discussion here: https://github.com/ros2/rmw_cyclonedds/pull/324#issuecomment-871570991. As mentioned at the bottom of that PR, LTTng has only partial support for tracing SDT probes, but it can generate them alongside its own tracepoints if it is built from source and configured with `--with-sdt`.

Note: this PR targets `master`, but I'd like to get it cherry-picked (if the PR is accepted) to the `iceoryx` branch so that ROS 2 can use it (since it currently targets the `iceoryx` branch).

Signed-off-by: Christophe Bedard <bedard.christophe@gmail.com>

### CODE REVIEW

**Review by eboasson — COMMENTED** [review]
Hi @christophebedard thanks for contributing this first attempt at introducing proper tracing into the Cyclone code base. I have a problem with basing everything exclusively on LTTng, as I really want to (also) be able to use `dtrace` on macOS, for example.

That means probe definitions need to be a bit differently. This in itself is going to be tricky, so I would really like to see a PoC for first, or at least a sketch of one 😁.

Based on what you noted in https://github.com/ros2/rmw_cyclonedds/pull/324#issuecomment-884568893 concerning semaphores, an argument could be made that LTTng tracepoints would necessarily end up in less interesting places and fewer in number and that consequently it wouldn't be worth the effort. In that case, arguably, LTTng tracepoints may not be a good choice for Cyclone to begin with.

That restriction does strike me as odd, as I haven't seen it in the LTTng documentation and instead found a remark that it uses a lock-free procedure for adding to the trace. I don't see how the two fit together.

Some of the comments are perhaps a bit stream-of-consciousness ... my apologies for that.

File: src/core/ddsc/CMakeLists.txt [hunk 1]
```diff
@@ -115,6 +120,10 @@ target_include_directories(ddsc
   INTERFACE
     $<INSTALL_INTERFACE:include>)
 
+if (DDS_HAS_LTTNG_TRACING)
+  target_link_libraries(ddsc PRIVATE ${LTTNG_LIBRARIES} "-rdynamic")
```
[the remark this answers is not in the document]
Comment (christophebedard) [reply src/core/ddsc/CMakeLists.txt:124]: I've double checked, and it's indeed not needed here.

It was needed in order to resolve function pointers to symbols in `rclcpp`, so it's not needed here, but I forgot about it when copying the CMake lines from `tracetools`.

**Comment by thijsmie** [comment]
While I would actually like to merge this PR it's been left too long and has become unreviewable with the many conflicts and highly diverged branches. @christophebedard would you like to rebase your changes onto the current master or should I close this and defer LTTng to the TODO backlog?

**Comment by christophebedard** [comment]
Thanks for the follow up. We can close this for now.

**Pull request closed without merging**